### PR TITLE
Document and bring sanity to BoxConstraints

### PR DESCRIPTION
**Pull request by abarth**
Turns out many of the functions on BoxConstraints weren't used or had callers
that could easily be updated to other functions. I've added dartdoc to all the
public functions as well as renamed some functions that had similar names but
did different things.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

File: examples/widgets/styled_text.dart [hunk 1]
```diff
@@ -53,7 +53,8 @@ HAL: This mission is too important for me to allow you to jeopardize it.''';
 
   Component createSeparator() {
     return new Container(
-      constraints: const BoxConstraints.expandWidth(maxHeight: 0.0),
+      constraints: const BoxConstraints.tightFor(
+        width: double.INFINITY, height: 0.0),
```
Comment (Hixie) [inline examples/widgets/styled_text.dart]: I think we should avoid having double.INFINITY in the call sites as much as possible, on the principle that infinities are confusing and it's an ugly constant. Keeping the use of infinity as an internal implementation detail feels nicer. As such, I think this particular one was better using a different constructor. Having said that, maybe BoxConstraints.expand(minHeight: 0.0, maxHeight: 0.0), rather than expandWidth(), would work, so we only need one expand() constructor?

File: sky/packages/sky/lib/src/rendering/box.dart [hunk 2]
```diff
@@ -23,7 +23,22 @@ class _DebugSize extends Size {
   final bool _canBeUsedByParent;
 }
 
+/// Immutable layout constraints for box layout
+///
+/// A size respects a BoxConstraints if, and only if, all of the following
+/// relations hold:
+///
+/// * `minWidth <= size.width && size.width <= constraints.maxWidth`
+/// * `minHeight <= size.height && size.height <= maxHeight`
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: why do you use `a < b && b < c` here, but just `a < b < c` below?

File: sky/packages/sky/lib/src/rendering/box.dart [hunk 3]
```diff
@@ -23,7 +23,22 @@ class _DebugSize extends Size {
...
+///
+/// The box constraints itself must satisfy these relations:
+///
+/// * `0.0 <= minWidth <= maxWidth <= double.INFINITY`
+/// * `0.0 <= minHeight <= maxHeight <= double.INFINITY`
+///
+/// Note: `double.INFINITY` is a legal value for each constraint.
 class BoxConstraints extends Constraints {
+  /// Constructs a box constraints with these constraints
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: `Constructs a BoxConstraints` or `Constructs box constraints`

File: sky/packages/sky/lib/src/rendering/box.dart [hunk 4]
```diff
@@ -23,7 +23,22 @@ class _DebugSize extends Size {
...
 
+/// Immutable layout constraints for box layout
+///
+/// A size respects a BoxConstraints if, and only if, all of the following
+/// relations hold:
+///
+/// * `minWidth <= size.width && size.width <= constraints.maxWidth`
+/// * `minHeight <= size.height && size.height <= maxHeight`
+///
+/// The box constraints itself must satisfy these relations:
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: `BoxConstraints itself` or `box constraints themselves`

File: sky/packages/sky/lib/src/rendering/box.dart [hunk 5]
```diff
@@ -31,12 +46,19 @@ class BoxConstraints extends Constraints {
     this.maxHeight: double.INFINITY
   });
 
+  final double minWidth;
+  final double maxWidth;
+  final double minHeight;
+  final double maxHeight;
+
+  /// Constructs a box constraints that is respected only by size
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: `by the given size` or find some way to make `size` typographically denote the argument, because otherwise that's confusing (it could mean, e.g., respected only by size but not by position?)

File: sky/packages/sky/lib/src/rendering/box.dart [hunk 6]
```diff
@@ -45,24 +67,14 @@ class BoxConstraints extends Constraints {
...
-      maxWidth = double.INFINITY,
-      minHeight = 0.0;
-
-  const BoxConstraints.expandHeight({
-    this.maxWidth: double.INFINITY
-  }): minWidth = 0.0,
-      minHeight = double.INFINITY,
-      maxHeight = double.INFINITY;
-
+  /// A box constraints that requires an infinite width and height
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: rather than defining this in terms of infinite width and height, explain it in terms of being expected to be constrained by another BoxConstraints, so as to select the maximum widths and heights of those constraints.

File: sky/packages/sky/lib/src/rendering/box.dart (hunk 3 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: ditto for subsequent comments. "box constraints" is plural, "BoxConstraints" is singular.

File: sky/packages/sky/lib/src/rendering/box.dart [hunk 7]
```diff
@@ -91,7 +107,8 @@ class BoxConstraints extends Constraints {
     );
   }
 
-  BoxConstraints apply(BoxConstraints constraints) {
+  /// Returns a new box constraints whose values respect the given constraints
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: they do more than just respect the given constraints. I mean, a copy constructor would satisfy that definition. :-) This method merges the two sets of BoxConstraints by limiting this object by the requirements described by the other, or something.

File: sky/packages/sky/lib/src/rendering/box.dart [hunk 8]
```diff
@@ -100,80 +117,65 @@ class BoxConstraints extends Constraints {
...
     return result;
   }
+
+  /// The biggest size that satisifes the constraints
   Size get biggest => new Size(constrainWidth(), constrainHeight());
+
+  /// The smallest size that satisfies the constraints
   Size get smallest => new Size(constrainWidth(0.0), constrainHeight(0.0));
 
+  /// Whether either the width or the height is unbounded
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/box.dart]: actually isInfinite only returns true if they're both infinite. And as far as I can tell it's only used in two asserts? Maybe we can drop this. It's confusing that it doesn't work quite like size.isInfinite.

**Comment by Hixie** [comment]
LGTM modulo comments

File: sky/packages/sky/lib/src/rendering/box.dart (hunk 8 above)
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/box.dart]: Yeah, given that it's only used in the sector example, I've removed it.

File: sky/packages/sky/lib/src/rendering/box.dart (hunk 7 above)
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/box.dart]: Yeah, it's a bit wordy to describe.  Fixed.

File: sky/packages/sky/lib/src/rendering/box.dart (hunk 6 above)
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/box.dart]: Done.

File: sky/packages/sky/lib/src/rendering/box.dart (hunk 5 above)
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/box.dart]: Fixed.

File: sky/packages/sky/lib/src/rendering/box.dart (hunk 2 above)
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/box.dart]: Fixed.  The source this came from had a more complex condition.

File: examples/widgets/styled_text.dart (hunk 1 above)
Comment (abarth) [inline examples/widgets/styled_text.dart]: Done.

**Comment by abarth** [comment]
I believe I've addressed all your comments.  I'll land it in a bit if you don't have more feedback.

Thanks for the review!
